onnxruntime/core/mlas/lib/reorder.cpp

```diff
@@ -396,103 +411,191 @@ Return Value:
+                ss += 4;
+                dd += 4 * OutputSize;
+            }
 
-                s += 4 * BlockSize;
-                d += 4;
+            for (; bc < OutputChannelsThisIteration; bc += 1) {
+                MlasReorderGatherFloat32x4(ss, dd, BlockSize);
+                ss += 1;
+                dd += OutputSize;
             }
 
-            for (; OutputSizeRemaining > 0; OutputSizeRemaining--) {
+            s += 4 * BlockSize;
+            d += 4;
+        }
 
-                const float* ss = s;
-                float* dd = d;
-                size_t bc = 0;
+        for (; OutputSizeRemaining > 0; OutputSizeRemaining--) {
 
-                for (; bc < AlignedOutputChannelsThisIteration; bc += 4) {
-                    MlasReorderScatterFloat32x4(ss, dd, OutputSize);
-                    ss += 4;
-                    dd += 4 * OutputSize;
-                }
+            const float* ss = s;
+            float* dd = d;
+            size_t bc = 0;
 
-                for (; bc < OutputChannelsThisIteration; bc += 1) {
-                    *dd = *ss++;
-                    dd += OutputSize;
-                }
+            for (; bc < AlignedOutputChannelsThisIteration; bc += 4) {
+                MlasReorderScatterFloat32x4(ss, dd, OutputSize);
+                ss += 4;
+                dd += 4 * OutputSize;
+            }
 
-                s += BlockSize;
-                d += 1;
+            for (; bc < OutputChannelsThisIteration; bc += 1) {
+                *dd = *ss++;
+                dd += OutputSize;
             }
 
-            S += BlockSize * OutputSize;
-            D += OutputChannelsThisIteration * OutputSize;
+            s += BlockSize;
+            d += 1;
         }
+
+        S += BlockSize * OutputSize;
+        D += OutputChannelsThisIteration * OutputSize;
     }
 }
 
+
+void
+MLASCALL
+MlasReorderOutputNchw(
+    const int64_t* OutputShape,
+    const float* S,
+    float* D,
+    MLAS_THREADPOOL* ThreadPool
+    )
+/*++
+
+Routine Description:
+
+    This routine reorders an output buffer from NCHWc to NCHW format.
+
+Arguments:
+
+    OutputShape - Supplies the shape of the output tensor.
+
+    S - Supplies the address of the source tensor.
+
+    D - Supplies the address of the destination tensor.
+
+Return Value:
+
+    None.
+
+--*/
+{
+    MLAS_REORDER_OUTPUT_NCHW_BLOCK WorkBlock;
+
+    //
+    // Capture the NCHW reorder output operation parameters to the work block.
+    //
+
+    WorkBlock.S = S;
+    WorkBlock.D = D;
+    
+    WorkBlock.OutputChannels = size_t(OutputShape[1]);
+    WorkBlock.OutputSize = size_t(OutputShape[2]) * size_t(OutputShape[3]);
+
+    const size_t BlockSize = MlasNchwcGetBlockSize();
+    const size_t TasksPerBatch = size_t(ceil(((float)WorkBlock.OutputChannels) / BlockSize));
+    const size_t BatchCount = size_t(OutputShape[0]);
+    const size_t TasksCount = BatchCount * TasksPerBatch;    
+    WorkBlock.TasksCount = TasksCount;
```

Comment:
Did you try just parallel on N(batch size)?

Reply:
I have not. While it is certainly possible to parallelize just on N, some of our use cases have N=1, and we will be missing out on the opportunity to speed up those workloads.